text/0250-Composite_Samplers.md

```diff
@@ -0,0 +1,308 @@
+
+- If any of the delegates returned a threshold value from the range of `0` to `2^56-1`, the resulting threshold is the minimum value from the set of results from within that range.
+- Otherwise, the result is obtained by calling `GetThreshold` on `ConsistentAlwaysOffSampler`.
+
+Each delegate sampler MUST be given a chance to participate in calculating the threshold as described above and MUST see the same argument values. The order of the delegate samplers does not matter.
+
+### ConsistentRateLimiting
+
+`ConsistentRateLimiting` is a composite sampler that helps control the average rate of sampled spans while allowing another sampler (the delegate) to provide sampling hints.
+
+#### Required Arguments for ConsistentRateLimiting
+
+- ConsistentProbabilitySampler (delegate)
+- maximum sampling (throughput) target rate
+
+The sampler SHOULD measure and keep the average rate of incoming spans, and therefore also of the desired ratio between the incoming span rate to the target span rate.
+Upon invocation of its `GetThreshold` function, the composite sampler MUST get the threshold from the delegate sampler, passing the same arguments as received.
+If using the obtained threshold value as the final threshold would entail sampling more spans than the declared target rate, the sampler SHOULD increase the threshold to a value that would meet the target rate. Several algorithms can be used for threshold adjustment, no particular behavior is prescried by the specification though.
+
+When using `ConsistentRateLimiting` in our requirements example as a replacement for `EachOf` and `RateLimiting`, we are left with no use case for any direct equivalent to `EachOf` in Approach Two.
+
+## Summary - Approach Two
+
+### Example - sampling configuration with Approach Two
+
+With the samplers introduced by Approach Two, our example requirements can be coded in a very similar way as with ApproachOne. However, the work of the samplers configured this way forms a tree of `GetThreshold` invocations rather than `ShouldSample` invocations as in ApproachOne.
+
+```
+S = ConsistentRateLimiting(
+     ConsistentAnyOf(
+       ConsistentParentBased(
+         ConsistentRuleBased(ROOT, {
+             (http.target == /healthcheck) => ConsistentAlwaysOff,
+             (http.target == /checkout) => ConsistentAlwaysOn,
+             true => ConsistentFixedThreshold(0.25)
+         }),
+       ConsistentRuleBased(CLIENT, {
+         (http.url == /foo) => ConsistentAlwaysOn
+       }
+     ),
+     1000 * 60
+   )
+```
+
+### Limitations of composite samplers in Approach Two
+
+While making sampling decisions with samplers from Approach Two is more efficient and avoids dealing with non-mainstream cases, it puts some limits on the capabilities of the Consistent Probability Samplers. In particular, a custom CP sampler that wishes to add a span `Attribute` or modify TraceState will be out of luck if it is used as a delegate.
```

Comment:
As it is explicitly mentioned as a limitation. I wonder why a consistent sampler would add any span attributes or change the trace state at all (other than `th` and `rv`)? What would be a use case for this?

Reply:
Good question. With trace state, I imagine a head sampler can _mark_ some traces for downstream (head) samplers and for further trace processing, effectively classifying traces according to arbitrary criteria. Obviously, this requires writing custom samplers, but these samplers can be still CP compliant.
Adding Attributes is similar, even though it works only on the Span level and only for sampled Spans.

Reply:
Got it. I think this could also be achieved with Approach Two by extending the ConsistentSampler interface by methods other than `getThreshold`such as `getAttributes` and defining some logic how those attributes are finally added in case the overall sampling decision is positive. So I would not necessarily see that as limitation of Approach Two as we also have to define some logic for the attributes for Approach One.

Reply:
Yes, and we could do a similar thing with `TraceState` by providing `adjustTraceState`. For that call, we could pass the final `SamplingDecision` (which can be different from what the delegate decided). One difficulty is that for `RuleBased`, the Predicates would get re-evaluated in order to route the call to the correct delegate.

Reply:
I am thinking in the same direction. I have a sandbox where I've begun drafting my hopes&dreams for a V2 sampler API, right now I have something like:

```
type SamplingResult struct {
	Decision SamplingDecision
	Mutation func(trace.TraceState) ([]attribute.KeyValue, trace.TraceState)
}
```

for an Approach-2 sampler, the idea is that a composition of sampler delegates will first evaluate the decision, then it would modify the special TraceState fields, then for the positive decision makers it would allow them (sequentially) to apply modified attributes and modify the trace state in other ways. I think it will be fine to say that attributes are applied in the same order as the original delegate list, and that later samplers clobber earlier samplers in case of conflicts. I will flush this out more, but I think we can get past the trouble associated with modifying attributes and tracestate independent of the decision & essential tracestate updates.

Reply:
I modified the document to present this two-phase process.

Reply:
I've begun to prototype https://github.com/jmacd/opentelemetry-go/pull/877